### DISCUSSION TITLE
Two. Communication and the Use of Language

### DISCUSSION THREAD

**Discussion by Mehdi-H** [Mehdi-H/DDD-blue-book-collaborative-reading#3] [Chapter]
# Two. Communication and the Use of Language

You can discuss here about this chapter ! 👇

Ask questions, or share anything you find insightful !

**Comment by sofcalca** [comment]
## 🚧 Glossary

UBIQUITOUS LANGUAGE = 
CONTEXT MAPS =
BOUNDED CONTEXTS =
MODEL-DRIVEN DESIGN =

## What I learned from this chapter

The domain has its jargon, developers have theirs. If they use them to communicate between them, they won't  understand each other.
- Translation is costly.
- Terminology misalignment is a source of misunderstanding.
    - An abstraction in code is not called the same way orally
    - The lack of a common terminology means different people talk about the same thing using different words.
    - This creates confusion and is a huge source of errors.

### Solution: ubiquitous language

= A common vocabulary between business and developers.

- Used to communicate with each other, but also among themselves.
- Used to name objects in the code.
- Teams should force themselves to use this language to detect terms that sound odd and find alternatives that make more sense
- If a term changes, it has to be used in conversations, in the model, in the code, in diagrams, in documentation, etc.

Evans insists on the need to **force ourselves to use the ubiquitous language every time we talk about the software and the associated system**, a bit like forcing ourselves to speak a foreign language during classes, by banning our mother tongue. New, simpler and idiomatic ways of saying things will be integrated into the language.

> In other words, a UML diagram cannot convey two of the most important aspects of a model: the meaning of the concepts it represents, and what the objects are meant to do.

### Documentation methods

**Diagrams**: useful as a support for discussion: they should remain simple and not exhaustive.

**Documentation**: Xtreme programming says that code is enough as documentation. Evans says that this level of details is too detailed to meet all documentation needs.

> Documents can clarify design intent when the programming language does not support a straightforward implementation of a concept. Written documents should complement the code and the talking.

—> This makes me think of ADRs.

> By keeping documents minimal and focusing them on *complementing*  code and conversation, documents can stay connected to the project. Let the UBIQUITOUS LANGUAGE  and its evolution be your guide to choosing documents that live and get woven into the project’s activity.

If the document is not updated frequently (e.g., as the ubiquitous language evolves), it is better to archive it.

_All quotes are from Eric Evan's Domain Driven Design book_

**Reply by Mehdi-H** [reply]
> ## 🚧 Glossary
> UBIQUITOUS LANGUAGE = 
> CONTEXT MAPS = 
> BOUNDED CONTEXTS = 
> MODEL-DRIVEN DESIGN =
> 

💡 What do you think of initiating a GLOSSARY.md file in this repo to have a document with definition of "DDD"-related terms ?

**Reply by Mehdi-H** [reply]
> If the document is not updated frequently (e.g., as the ubiquitous language evolves), it is better to archive it.

This makes me think about the Living Documentation[^1] book which proposes mainly two strategies :

- generate some documentation artifacts from the code (like a glossary if you put enough effort to model business concepts in your source code), to keep code (the source of truth) and documentation in sync,
- maintain what Cyrille Martraire calls an _Evergreen document_ with high-level statements, product goals and vision : 

> Evergreen content is a kind of content that remains of interest for a long period of time, without change, for a particular audience. Evergreen content does not change, and yet it remains useful, relevant, and accurate.[^1]

[^1]: Martraire, Cyrille, _[Living Documentation: Continuous Knowledge Sharing by Design, First Edition](https://learning.oreilly.com/library/view/living-documentation-continuous/9780134689418/)_, Addison-Wesley Professional, 2019 May.

**Reply by sofcalca** [reply]
> :bulb: What do you think of initiating a GLOSSARY.md file in this repo to have a document with definition of "DDD"-related terms ?

Great idea. I made a [pull request](https://github.com/Mehdi-H/DDD-blue-book-collaborative-reading/pull/6) following your suggestion.

**Reply by Mehdi-H** [reply]
ℹ️ @sofcalca : I opened this issue #7 to follow-up on this glossary init' 👀 
Edit: issue #7 is closed ✅
